### PR TITLE
fix: controlling issues

### PR DESCRIPTION
**Pull request by rangoo94**
## Checklist (choose whats happened)

- [ ] breaking change! (describe)
- [x] tested locally
- [x] tested on cluster
- [ ] added new dependencies
- [ ] updated the docs
- [ ] added a test

## Breaking changes

-

## Changes

-

## Fixes

-

### CODE REVIEW

File: pkg/testworkflows/testworkflowcontroller/controller.go [hunk 1]
```diff
@@ -245,20 +246,25 @@ func (c *controller) WatchLightweight(parentCtx context.Context) <-chan Lightwei
...
 					status = testkube.QUEUED_TestWorkflowStatus
 				}
 				current = v.Value.Result.Current(sig)
+				isFinished = v.Value.Result.IsFinished()
 			}
 
-			if nodeName != prevNodeName || podIP != prevPodIP || prevStatus != status || prevCurrent != current {
+			// TODO: the final status should always have the finishedAt too,
+			//       there should be no need for checking isFinished diff
+			if nodeName != prevNodeName || isFinished != prevIsFinished || podIP != prevPodIP || prevStatus != status || prevCurrent != current {
```
Comment (vsukhin) [inline pkg/testworkflows/testworkflowcontroller/controller.go:262]: such code will not be simple, to maintain (

File: pkg/testworkflows/testworkflowcontroller/watchinstrumentedpod.go [hunk 2]
```diff
@@ -65,13 +66,27 @@ func WatchInstrumentedPod(parentCtx context.Context, clientSet kubernetes.Interf
...
-		podObj := <-pod.Peek(ctx)
+		var podObj *corev1.Pod
+		select {
+		// Obtain the Pod information for further execution
+		case p := <-pod.Peek(ctx):
+			podObj = p
+		// Handle when the execution have been finished, but there may be no Pod
+		case <-state.Finished(""):
+			select {
+			case <-time.After(1 * time.Second):
```
Comment (vsukhin) [inline pkg/testworkflows/testworkflowcontroller/watchinstrumentedpod.go:83]: would be nice to use const

File: pkg/testworkflows/testworkflowcontroller/podstate.go [hunk 3]
```diff
@@ -381,58 +411,97 @@ func (p *podState) ContainerResult(name string) (ContainerResult, error) {
...
+				// Try to firstly finish with the Pod information when it's possible
+				if IsJobDone(v.Value) && state.FinishedAt("").IsZero() && HadPodScheduled(v.Value) {
+					select {
+					case p, ok := <-podCh:
+						if p.Error != nil {
+							errorHandler(p.Error)
+						} else if ok {
+							state.RegisterPod(p.Value)
+						}
+					case <-time.After(2 * time.Second):
```
Comment (vsukhin) [inline pkg/testworkflows/testworkflowcontroller/podstate.go:498]: some const value

**Review by vsukhin — APPROVED** [review]

File: pkg/testworkflows/testworkflowcontroller/controller.go (hunk 1 above)
Comment (rangoo94) [reply pkg/testworkflows/testworkflowcontroller/controller.go:262]: I fully agree, although this code is rather not meant to be maintained. This PR, along actual bugfixes, is meant to "enable" the new orchestration with similar watching system.

The (mainly) `WatchInstrumentedPod` and `TestWorkflowResult` have so many edge cases handled, clock calibration, and auto-healing mechanisms implemented, that at this point it's probably better to just rewrite them based on the observations that we have for the last few months of running Test Workflows. After that, we will likely not need conditions like these at all.

**Pull request merged**